src/hotspot/cpu/x86/macroAssembler_x86.cpp

```diff
@@ -5162,7 +5162,24 @@ void MacroAssembler::restore_cpu_control_state_after_jni(Register rscratch) {
   // or verify that it wasn't changed (with -Xcheck:jni flag).
   if (VM_Version::supports_sse()) {
     if (RestoreMXCSROnJNICalls) {
+      Label FAIL, DONE;
+#ifdef _LP64
+      // Perform a little arithmetic to make sure that denormal
+      // numbers are handled correctly, i.e. that the "Denormals Are
+      // Zeros" flag has not been set.
```

Comment:
I don't understand what this part is doing. I thought it was simply checking so you could log/warn if the unexpected mode was detected. But it seems to cause MXCSR to not be restored when there is an issue, where I would expect you would always want to restore to overwrite the invalid mode the JNI call made. ??

Reply:
If we reach `FAIL`, MXCSR is reloaded from `addr_mxcsr_std()`, restoring correct IEEE behaviour. That's what RestoreMXCSROnJNICalls is supposed to do, as far as I can tell.

But I will take this part out, because there are other flags in MXCSR, so there is a potential compatibility problem if e.g. the Precision Mask were set in a JNI call.

Reply:
Misread the code flow. So you were skipping the restore if you didn't think it was needed? The fact the restore was now conditional is what threw me.